chain/src/main/scala/org/bitcoins/chain/models/CompactFilterHeaderDAO.scala

```diff
@@ -124,7 +124,11 @@ case class CompactFilterHeaderDAO()(implicit
   private val bestFilterHeaderQuery = {
     val join = table
       .join(blockHeaderTable)
-      .on(_.blockHash === _.hash)
+      //sort by height as an optimization
+      .sortBy(_._1.height.desc)
+      //just take the last 2016 headers, if we have a reorg larger than
+      //this we will not be able to retrieve that header
+      .take(appConfig.chain.difficultyChangeInterval)
```

Comment:
This should make it slower, it should be faster if we do what we are doing for filters:

```
.map(_.headOption.map(_.height))
 ```

Reply:
However, these seem like they should resolve to the same sql query

Reply:
That will not reduce the set size, the point of the `.take()` is to make sure that the subsequent [filter](https://github.com/christewart/bitcoin-s-core/blob/5ce5b71250ec184c8348517a8fe36ca3ba351e51/chain/src/main/scala/org/bitcoins/chain/models/CompactFilterHeaderDAO.scala#L136) call is not done over the _entire_ table. Only 2016 headers. 


Breaking down the calls 

1. `.sortBy()` is O(n) (we need to look through the entire table to establish an ordering of heights
2.  [`.max`](https://github.com/christewart/bitcoin-s-core/blob/5ce5b71250ec184c8348517a8fe36ca3ba351e51/chain/src/main/scala/org/bitcoins/chain/models/CompactFilterHeaderDAO.scala#L133) is O(n).
3. [filter](https://github.com/christewart/bitcoin-s-core/blob/5ce5b71250ec184c8348517a8fe36ca3ba351e51/chain/src/main/scala/org/bitcoins/chain/models/CompactFilterHeaderDAO.scala#L136) is also O(n)

If we reduce set size from ~2 million testnet headers, to just 2016 testnet headers, we should see significant performance improvements even if the algorithm is still `O(n^3)` 

This same methodology is applied to the filter table as well

Reply:
Because of our indexes it shouldn't be O(n) it should be O(log n)

Reply:
Also this removes the safety of checking by chain work